tensorboard/summary/_output.py

```diff
@@ -0,0 +1,114 @@
+
+from tensorboard.compat.proto import event_pb2
+from tensorboard.compat.proto import summary_pb2
+from tensorboard.summary.writer import event_file_writer
+
+import abc
+
+
+class Output(abc.ABC):
+    """Interface for emitting tensor-formatted summary data.
+
+    Implementations of this interface can be passed to Writer to customize
+    how summary data is actually persisted (e.g. to disk, to memory, over
+    the network, etc.).
+    """
+
+    @abc.abstractmethod
+    def emit_scalar(
+        self,
+        *,
+        plugin_name,
+        tag,
+        data,
+        step,
+        wall_time,
+        tag_metadata=None,
+        description=None,
+    ):
+        """Emits one scalar data point to this Output.
+
+        Args:
+          plugin_name: string name to uniquely identify the type of time series
+            (historically associated with a TensorBoard plugin).
+          tag: string tag used to uniquely identify this time series.
+          data: `np.float32` scalar value for this data point.
+          step: `np.int64` scalar step value for this data point.
+          wall_time: `float` seconds since the Unix epoch, representing the
+            real-world timestamp for this data point.
+          tag_metadata: optional bytes containing metadata for this entire time
+            series. This should be constant for a given tag; only the first
+            value encountered will be used.
+          description: optional string description for this entire time series.
+            This should be constant for a given tag; only the first value
+            encountered will be used.
+        """
+        pass
+
+    @abc.abstractmethod
+    def flush(self):
+        """Flushes any data that has been buffered."""
+        pass
+
+    @abc.abstractmethod
+    def close(self):
+        """Closes the Output and also flushes any buffered data."""
+        pass
+
+
+class DirectoryOutput(Output):
+    """Outputs summary data by writing event files to a log directory."""
+
+    def __init__(self, path):
+        """Creates a `DirectoryOutput` for the given path."""
+        self._ev_writer = event_file_writer.EventFileWriter(path)
+
+    def emit_scalar(
+        self,
+        plugin_name,
+        tag,
+        data,
+        step,
+        wall_time,
+        tag_metadata=None,
+        description=None,
+    ):
+        """See `Output`."""
+        # TODO(nfelt): cache summary metadata to emit only once.
+        summary_metadata = summary_pb2.SummaryMetadata(
+            plugin_data=summary_pb2.SummaryMetadata.PluginData(
+                plugin_name=plugin_name, content=tag_metadata
+            ),
+            summary_description=description,
+            data_class=summary_pb2.DataClass.DATA_CLASS_SCALAR,
+        )
+        event = event_pb2.Event(wall_time=wall_time, step=step)
+        event.summary.value.add(
+            tag=tag, simple_value=data, metadata=summary_metadata
```

Comment:
With `DATA_CLASS_SCALAR`, [time series are expected to have `tensor`
values rather than `simple_value`s][spec]. I think technically both
RustBoard and the Python reading logic will happen to accept your data,
but it would be nice to conform to spec. Should we change this to emit
tensors, or do you think that we should make the spec more permissive?

[spec]: https://github.com/tensorflow/tensorboard/blob/742781d94413533df799143772dd6fd4bdc5091a/tensorboard/compat/proto/summary.proto#L70-L73